test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,271 @@
+            throw new RuntimeException(error);
+        }
+    }
+
+    private static void withSocketImplFactory(SocketImplFactory newFactory, ThrowingRunnable runnable) throws Throwable {
+        SocketImplFactory oldFactory = (SocketImplFactory) SOCKET_IMPL_FACTORY_HANDLE.getAndSet(newFactory);
+        try {
+            runnable.run();
+        } finally {
+            SOCKET_IMPL_FACTORY_HANDLE.set(oldFactory);
+        }
+    }
+
+    private static final class MockSocketImpl extends ThrowingSocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindError;
+
+        private final Exception connectError;
+
+        private MockSocketImpl(Exception bindError, Exception connectError) {
+            this.bindError = bindError;
+            this.connectError = connectError;
+        }
+
+        @Override
+        protected void create(boolean stream) {}
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindError);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectError);
+        }
+
+        private void throwIfPresent(Exception connectError) throws IOException {
+            if (connectError != null) {
+                switch (connectError) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown connectError type: " + connectError.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("ctor_should_close_on_failures")
+    @SuppressWarnings("resource")
+    void ctor_should_close_on_failures(TestCase testCase) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        withSocketImplFactory(() -> testCase.socketImpl, () -> {
+
+            // Trigger the failure
+            Exception error = assertThrows(Exception.class, () -> {
+                InetAddress address = InetAddress.getLoopbackAddress();
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                new Socket(address, 0xDEAD, address, 0xBEEF);
+            });
+
+            // Run verifications
+            testCase.caughtErrorVerifier.accept(error);
+            testCase.socketImplVerifier.run();
+
+        });
+
+    }
+
```

Comment:
static List<TestCase> ctorCloseOnFailuresTestCases() {
         return List.of(
                 TestCase.BindFailureFactory.ioExceptionTestCase(),
                 TestCase.ConnectFailureFactory.ioExceptionTestCase(),
                 TestCase.ConnectFailureFactory.illegalArgumentExceptionTestCase(1));
     }

c.f. name change suggestion below

Reply:
Applied this suggestion in e37ee840b9681aa017f8653df68c69232a6b6c83.